### PR TITLE
feat: [PL-43269]: Enhancement on Move Project API

### PR DESCRIPTION
**Pull request by adiyaar24**
## Describe your changes
This PR adds the ProjectMoveEvent and shows diff in Audit Trails

Screenshot 2023-12-18 at 3 53 03 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56894)

### CODE REVIEW

**Comment by adiyaar24** [comment]
trigger uts

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56894)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkMetadataConstants.java` line 29 at r1](https://reviewable.io/reviews/harness/harness-core/56894#-NlwZ9Cd7YMl2jPk8QlD:-NlwZ9Cd7YMl2jPk8QlE:b8srgt8) ([raw file](https://github.com/harness/harness-core/blob/03328aece91a9a1d5bbe4587568fea18f6f7586e/953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkMetadataConstants.java#L29)):*
> ```Java
>   public static final String RESTORE_ACTION = "restore";
>   public static final String UPDATE_ACTION = "update";
>   public static final String MOVE_ACTION = "moved";
> ```

move not moved
___
*[`platform-service/modules/audit-service/contracts/src/main/java/io/harness/audit/Action.java` line 53 at r1](https://reviewable.io/reviews/harness/harness-core/56894#-NlwZCzM9Ods3QMQirtP:-NlwZCzM9Ods3QMQirtQ:bb6zpli) ([raw file](https://github.com/harness/harness-core/blob/03328aece91a9a1d5bbe4587568fea18f6f7586e/platform-service/modules/audit-service/contracts/src/main/java/io/harness/audit/Action.java#L53)):*
> ```Java
>   ROLE_ASSIGNMENT_DELETED,
>   MOVED
> }
> ```

MOVE

**Review by ankit-harness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56894)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 511 at r1](https://reviewable.io/reviews/harness/harness-core/56894#-NlwalVs0L6O3bSLZgOW:-NlwalVs0L6O3bSLZgOX:b-r00v6a) ([raw file](https://github.com/harness/harness-core/blob/03328aece91a9a1d5bbe4587568fea18f6f7586e/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L511)):*
> ```Java
>     if (optionalProject.isPresent()) {
>       Project project = optionalProject.get();
>       Project oldProject = (Project) HObjectMapper.clone(project);
> ```

Any side effects of using HObjectMapper?

**Review by adiyaar24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56894)* status: 7 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 511 at r1](https://reviewable.io/reviews/harness/harness-core/56894#-NlwalVs0L6O3bSLZgOW:-NlyGQCLCNaiDI73QRLe:b32w12) ([raw file](https://github.com/harness/harness-core/blob/03328aece91a9a1d5bbe4587568fea18f6f7586e/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L511)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Any side effects of using HObjectMapper?
</blockquote></details>

No
___
*[`953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkMetadataConstants.java` line 29 at r1](https://reviewable.io/reviews/harness/harness-core/56894#-NlwZ9Cd7YMl2jPk8QlD:-NlyGS-s6tzYVsLPy-_M:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/03328aece91a9a1d5bbe4587568fea18f6f7586e/953-events-api/src/main/java/io/harness/eventsframework/EventsFrameworkMetadataConstants.java#L29)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

move not moved
</blockquote></details>

Done.
___
*[`platform-service/modules/audit-service/contracts/src/main/java/io/harness/audit/Action.java` line 53 at r1](https://reviewable.io/reviews/harness/harness-core/56894#-NlwZCzM9Ods3QMQirtP:-NlyGUB5--L7KtTHLr-1:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/03328aece91a9a1d5bbe4587568fea18f6f7586e/platform-service/modules/audit-service/contracts/src/main/java/io/harness/audit/Action.java#L53)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

MOVE
</blockquote></details>

Done.

**Comment by adiyaar24** [comment]
trigger uts

**Comment by adiyaar24** [comment]
trigger uts

**Comment by adiyaar24** [comment]
trigger ut7, ut8

File: 120-ng-manager/src/main/java/io/harness/ng/core/events/ProjectMoveEvent.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+public class ProjectMoveEvent implements Event {
+  private String accountIdentifier;
+  ProjectDTO newProject;
+  ProjectDTO oldProject;
+
+  public ProjectMoveEvent(String accountIdentifier, ProjectDTO newProject, ProjectDTO oldProject) {
+    this.newProject = newProject;
+    this.oldProject = oldProject;
+    this.accountIdentifier = accountIdentifier;
+  }
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/events/ProjectMoveEvent.java:38]: nit: can be covered by `@AllArgsContructor`

File: 120-ng-manager/src/main/java/io/harness/ng/core/event/ConnectorEntityCRUDStreamListener.java [hunk 2]
```diff
@@ -190,4 +193,19 @@ private boolean processProjectDeleteEvent(ProjectEntityChangeDTO projectEntityCh
...
     return true;
   }
+
+  private boolean processProjectMoveEvent(ProjectEntityChangeDTO projectEntityChangeDTO) {
+    String accountIdentifier = projectEntityChangeDTO.getAccountIdentifier();
+    Boolean isBuiltInSMDisabled =
+        parseBoolean(NGRestUtils
+                         .getResponse(settingsClient.getSetting(
+                             SettingIdentifiers.DISABLE_HARNESS_BUILT_IN_SECRET_MANAGER, accountIdentifier, null, null))
+                         .getValue());
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/event/ConnectorEntityCRUDStreamListener.java:203]: please handle exception case on this, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/outbox/ProjectEventHandler.java [hunk 3]
```diff
@@ -69,6 +70,8 @@ public boolean handle(OutboxEvent outboxEvent) {
           return handleProjectDeleteEvent(outboxEvent);
         case "ProjectRestored":
           return handleProjectRestoreEvent(outboxEvent);
+        case "ProjectMoved":
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/outbox/ProjectEventHandler.java:73]: nit: we should extract this in a constant, thanks

**Comment by pbarapatre10** [comment]
trigger uts

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r2, 1 of 2 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56894)* status: 7 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/outbox/ProjectEventHandler.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/56894#-NnClOKm06Z3uAkXzcBS:-NnClOKm06Z3uAkXzcBT:b-upv48p) ([raw file](https://github.com/harness/harness-core/blob/ed332ada6dcb2bbbd2b9a2298bfe58b666540aee/120-ng-manager/src/main/java/io/harness/ng/core/outbox/ProjectEventHandler.java#L74)):*
> ```Java
>         case "ProjectRestored":
>           return handleProjectRestoreEvent(outboxEvent);
>         case PROJECT_MOVED:
> ```

Should we be consistent here? I agree these events type strings can be converted into a constant. But, this is also bringing in inconsistency. Maybe we can separate out constants as a separate PR

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/outbox/ProjectEventHandler.java` line 74 at r4](https://reviewable.io/reviews/harness/harness-core/56894#-NnClOKm06Z3uAkXzcBS:-NnCtnyU5yRPpcWxrs5k:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/ed332ada6dcb2bbbd2b9a2298bfe58b666540aee/120-ng-manager/src/main/java/io/harness/ng/core/outbox/ProjectEventHandler.java#L74)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Should we be consistent here? I agree these events type strings can be converted into a constant. But, this is also bringing in inconsistency. Maybe we can separate out constants as a separate PR
</blockquote></details>

Done

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56894#-:-NnD4Pkq2QkIbCkahWjj:bnfp4nl)* status: 7 file(s) to review. 1 of 2 approvals obtained.

**Comment by pbarapatre10** [comment]
trigger uts

**Comment by pbarapatre10** [comment]
trigger ut6

**Comment by pbarapatre10** [comment]
trigger ut8

**Comment by pbarapatre10** [comment]
trigger ut9

**Comment by pbarapatre10** [comment]
trigger uts

**Comment by pbarapatre10** [comment]
trigger ut7

**Comment by pbarapatre10** [comment]
trigger ut9

**Pull request merged**